api/app/c_haines/fetch.py

```diff
@@ -0,0 +1,273 @@
+""" Fetch c-haines geojson
+"""
+from io import StringIO
+from typing import Final, Iterator
+from urllib.parse import urljoin
+from datetime import datetime, timedelta
+import logging
+from app import config
+import app.db.database
+from app.db.models.c_haines import SeverityEnum
+from app.schemas.weather_models import CHainesModelRuns, CHainesModelRunPredictions, WeatherPredictionModel
+from app.weather_models import ModelEnum
+from app.db.crud.c_haines import (get_model_run_predictions, get_most_recent_model_run,
+                                  get_prediction_geojson, get_prediction_kml, get_model_run_kml)
+
+logger = logging.getLogger(__name__)
+
+FOLDER_OPEN: Final = '<Folder>'
+FOLDER_CLOSE: Final = '</Folder>'
+
+
+# Severity enum -> Text description mapping
+severity_text_map = {
+    SeverityEnum.LOW: '<4 Low',
+    SeverityEnum.MODERATE: '4 - 8 Moderate',
+    SeverityEnum.HIGH: '8 - 11 High',
+    SeverityEnum.EXTREME: '11+ Extreme'
+}
+
+# Severity enum -> KML style mapping
+severity_style_map = {
+    SeverityEnum.LOW: 'low',
+    SeverityEnum.MODERATE: 'moderate',
+    SeverityEnum.HIGH: 'high',
+    SeverityEnum.EXTREME: 'extreme'
+}
+
+
+def get_severity_style(c_haines_index: SeverityEnum) -> str:
+    """ Based on the index range, return a style string """
+    return severity_style_map[c_haines_index]
+
+
+def open_placemark(model: ModelEnum, severity: SeverityEnum, timestamp: datetime):
+    """ Open kml <Placemark> tag. """
+    kml = []
+    kml.append('<Placemark>')
+
+    if model == ModelEnum.GDPS:
+        end = timestamp + timedelta(hours=3)
+    else:
+        end = timestamp + timedelta(hours=1)
+    kml.append('<TimeSpan>')
+    kml.append('<begin>{}</begin>'.format(timestamp.isoformat()))
+    kml.append('<end>{}</end>'.format(end.isoformat()))
+    kml.append('</TimeSpan>')
+    kml.append('<styleUrl>#{}</styleUrl>'.format(get_severity_style(severity)))
+    kml.append('<name>{}</name>'.format(severity_text_map[severity]))
+    kml.append('<MultiGeometry>')
+    return "\n".join(kml)
+
+
+def close_placemark():
+    """ Close kml </Placemark> tag. """
+    kml = []
+    kml.append('</MultiGeometry>')
+    kml.append('</Placemark>')
+    return "\n".join(kml)
+
+
+def add_style(kml, style_id, color):
+    """ Add kml <Style> tag """
+    kml.append('<Style id="{}">'.format(style_id))
+    kml.append('<LineStyle>')
+    kml.append('<width>1.5</width>')
+    kml.append('<color>{}</color>'.format(color))
+    kml.append('</LineStyle>')
+    kml.append('<PolyStyle>')
+    kml.append('<color>{}</color>'.format(color))
+    kml.append('</PolyStyle>')
+    kml.append('</Style>')
```

Comment:
Oof. Writing HTML in python one line at a time. That's rough.

Reply:
My initial thought was to use some kml or xml library - but doing it this way is considerably faster since having a streaming response is much easier to do this way.